### ISSUE TITLE
support audio recording

### ISSUE THREAD

**Issue by jodeleeuw** [jspsych/jsPsych#494]
this should be possible, but will require permissions from the user.

**Comment by mjaquiery** [comment]
I'm working on this.

**Comment by becky-gilbert** [comment]
In case anyone's interested: I've recently built on @mjaquiery's version of an audio recording plugin. My version can be found here: 
https://github.com/becky-gilbert/jsPsych/blob/audio-response/plugins/jspsych-image-audio-response.js 
https://github.com/becky-gilbert/jsPsych/blob/audio-response/examples/jspsych-image-audio-response.html
I'd love to get feedback from anyone who wants to try it out. Once we're happy with it, I'll create similar html-audio-response and audio-audio-response plugins, and submit a pull request along with examples and docs.

**Comment by hesmall** [comment]
If anyone is interested, I worked off of Becky's plugin and made an html-audio-response plugin that also has a parameter to allow the mic permissions popup (thanks to the audio recording code here: [https://experiments.ppls.ed.ac.uk/](url) ) and an option to manually end the recording by a key press (you can also specify which key this should be). 
That is here: [https://github.com/hesmall/custom_jsPsych_plugins/blob/master/jspsych-html-audio-response.js](url)

I also made an audio-audio-response plugin. It records audio after the stimulus audio has finished playing. It has the same option to manually end the recording and specify a key for that. 
That is here: [https://github.com/hesmall/custom_jsPsych_plugins/blob/master/jspsych-audio-audio-response.js](url)

**Comment by becky-gilbert** [comment]
Awesome, thanks @hesmall! It'd be great if we could combine these and then set up a pull request for the main library. I'll have a look at this as soon as I get the chance. Thanks again!

**Comment by becky-gilbert** [comment]
@hesmall thanks so much for these improvements! Would you mind submitting a pull request into my audio response branch? https://github.com/becky-gilbert/jsPsych/tree/audio-response

**Comment by mchlwgnr** [comment]
@hesmall @becky-gilbert thanks for making these plugins available! Do either of you have a working example audio-audio-response? I get an error for that one (the other two plugins work for me): 'Uncaught (in promise) TypeError: Cannot set property 'onended' of undefined  undefined at start_trial'. I think it's just not loading the soundfile, but I'm not sure why. Here's the code I use to call the plugin ((the .html is located in the examples folder, so the relative path should work? I'm running this in chrome, and I added a button click trial before to avoid audio being blocked): 

`timeline.push({
        type: 'audio-audio-response',
        audio_stimulus: 'sound/speech_red.mp3',
        visual_stimulus: 'stimulus goes here',
        prompt: 'prompt',
        allow_playback: true,
        manually_end_recording: true
    });`

**Comment by hesmall** [comment]
Hi @mchlwgnr!

I think this is because I didn't include this line in the plugin (which enables the automatic preloading):
jsPsych.pluginAPI.registerPreload('audio-audio-response', 'audio_stimulus', 'audio');

Manually preloading the audio stimuli works -- I tried your code out and got the same error, but when I preloaded the audio, it worked! 

However, I fixed the plugin to allow automatic preloading-- here! https://github.com/hesmall/custom_jsPsych_plugins 
So that code should work without manually preloading the stimuli!

**Comment by mchlwgnr** [comment]
Thanks for the quick response @hesmall! I added that already, but  I think
there's another problem: The script  can't find the variable 'audio' at
some point, and maybe there's variable-naming issue? Once I assign the
value of 'source' also to 'audio'  in the first part of the following
if-statement, then the script works:

        var context = jsPsych.pluginAPI.audioContext();
        var  audio = [];
        if(context !== null){
          var source = context.createBufferSource();
          source.buffer =
jsPsych.pluginAPI.getAudioBuffer(trial.audio_stimulus);
          source.connect(context.destination);
          audio = source;
        } else {
          audio = jsPsych.pluginAPI.getAudioBuffer(trial.audio_stimulus);
          audio.currentTime = 0;
        }

I'm not sure though whether doing this messes things up in some other way
(it *appears* to be working, but I haven't figured out yet all the uses of
'audio' and 'source' to  see whether i can just replace one with the other
throughout

**Comment by hesmall** [comment]
@mchlwgnr huh, well I'm glad you got it to work! When I use the updated audio-audio-response plugin, I don't have any issues, so I can't replicate your problem unfortunately. That if-statement came directly from the other audio plugins already included in jspsych, so I'm not sure what went wrong here! 

That fix seems like it wouldn't affect anything negatively, maybe @becky-gilbert can weigh in on that!

**Comment by becky-gilbert** [comment]
Hi @mchlwgnr and @hesmall! I'm not sure yet because I haven't had time to look at this in detail, but I think the problem with the audio-audio-response plugin might be on [line 240](https://github.com/hesmall/custom_jsPsych_plugins/blob/master/jspsych-audio-audio-response.js#L240). This line is setting the 'onended' function for the `audio` element, but the `audio` element only exists as a fallback when the browser doesn't support web audio, or when `use_webaudio` is set to `false` in `jsPsych.init` (see [line 188](https://github.com/hesmall/custom_jsPsych_plugins/blob/master/jspsych-audio-audio-response.js#L188) for how the `audio` element is first defined). @mchlwgnr's solution might be fine, but I think a better solution would be to set the audio's onended function inside of an if statement, like this one on [line 232](https://github.com/hesmall/custom_jsPsych_plugins/blob/master/jspsych-audio-audio-response.js#L232), so that the event listener can be attached to either the `audio` or `source` element as appropriate. 

I'll have a closer look at this as soon as I have time. In the meantime I hope this helps.

**Comment by mchlwgnr** [comment]
Thanks @becky-gilbert, that seems right:  the script works for me if I replace lines 240-245 with the following code  that also solves the problem (assuming we also add `jsPsych.pluginAPI.registerPreload('audio-audio-response', 'audio_stimulus', 'audio');` to the plugin). I can't say I fully understand the code yet, but just reporting how to get it to (at least apparently) work properly:

```     if(context !== null){
              source.onended = function() {
                start_time = performance.now(); //reset start time to measure rt from end of audio file
                if (trial.wait_for_mic_approval===false) {
                    start_recording();
                }
              }
            } else {
                audio.onended = function() {
                 start_time = performance.now(); //reset start time to measure rt from end of audio file
                 if (trial.wait_for_mic_approval===false) {
                    start_recording();
                 }
                }
            }

```

**Comment by mchlwgnr** [comment]
Can anyone point me to a worker script that I can use with this plugin to save the recording as a .wav or .mp3 file?

**Comment by jackdolgin** [comment]
Not to steer the conversation away but I've been trying to record audio for an online task and before encountering this thread I've been using the following script from @andrewheusser, which is also jspsych compatible I think. Just wanted to throw that out in case it's helpful!
https://github.com/ContextLab/autoFR/blob/master/exp/autoFR/static/lib/jspsych-5.0/plugins/jspsych-free-recall.js

**Comment by becky-gilbert** [comment]
@mchlwgnr I'm afraid not, as so far I've just been saving the data as a base 64 string (the plugin's default). How are you hosting your study? If you're a JATOS user you could use the [jatos.uploadResultFile](http://www.jatos.org/jatos.js-Reference.html#jatosuploadresultfile) function. Otherwise I'm hoping someone else can jump in to help!

@jackdolgin Thanks, that definitely sounds helpful. I'll check it out!

**Comment by lbinding** [comment]
Hi all, absolutely brilliant work! I'm very new to this so please excuse if this is a silly question, is it possible to only allow mic permissions once? I find it quite distracting to have to enable mic permissions each trial ~ thanks in advance! :)

**Comment by becky-gilbert** [comment]
@lbinding unfortunately this is a browser thing that I don't think we have control over. Are you running your task on a server, or opening the file directly in the browser? When I run an HTML file directly in the Chrome then it asks me for mic permission every time. But once the same file is hosted on a server, I'm only asked for permission on the first trial.

**Comment by mamuerie** [comment]
Hi all, first, thanks for your brilliant work on jspsych and on the increasing number of plugins! We also implemented the possibility to record short (1,5 sec) audio snippets and upload them after every trial to a jatos server as a wav file. We used parts from the following link: https://github.com/gilenta/jspsych-voice-recorder. After a few participants (all of them were using Chrome), we checked our wav files and realized that they differ in the number of recorded data points. The first recorded file has around 65536 data points and the last file only 20480, while the time for recording is constant and the sampling rate saved with the wav file also stays the same. Thus, it seems like that with time (the whole experiment lasts around 2,5 h and has more than 2000 wav files) some datapoints got lost on the way. Does anyone have an idea why this is happening and how we could prevent it? I thought it could be a performance issue like on youtube: performance is automatically decreasing as soon as the internet connection isn't fast enough. Does that makes sense and is there any way to control it? Any ideas are much appreciated! Thanks in advance!

**Comment by becky-gilbert** [comment]
Hi @mamuerie, I haven't used the voice recorder plugin that you're referring to and unfortunately I'm not sure what the problem is. But I did see in the [README notes](https://github.com/gilenta/jspsych-voice-recorder#notes) that this is a known issue, even in Chrome:

> The plugin currently only works on Chrome. It can technically run on Firefox and Safari, too, but if more recordings are taken in sequence these browser are likely to crash and / or not save the the recordings in their entirety. Even is Chrome is used, audio quality may start to deterioriate after a long series of recording taken in sequence (e.g. over 100 snippets of 6 seconds each).

You could try using one of our *-audio-response plugins that is currently still in development, like this html-audio-response plugin: https://github.com/becky-gilbert/jsPsych/blob/audio-response/plugins/jspsych-html-audio-response.js Of course you might have the same problem with this one, but so far I haven't had any reports yet of data loss or crashing from the people who have used it.

**Comment by mamuerie** [comment]
Hi @becky-gilbert thanks for your fast response! In this case, we will definitely try to switch to the html-audio-response! Thanks for pointing out the README notes!

**Comment by mamuerie** [comment]
Hi it's me again with another question. We successfully implemented the suggested plugin but now I am struggling with the resulting audio files. We're saving them as base64 file on our server and when I'm trying to decode it as a .wav file (using those suggestions: https://stackoverflow.com/questions/50279380/how-to-decode-base64-string-directly-to-binary-audio-format) the file is either too short and does not contain the response (using the wav.setparams part, resulting format: wav and pcm) or I'm able to hear the response when I listen to it via VLC Player but could not read the wav file into any other program like matlab or python ("error: unknown format or no riff", resulting format: WebM and Opus, when I use the simple version: file.write to a .wav file). 
How do you usually process the resulting audio files from the plugin and in which format do you save and transform them with what kind of tool/programming language? Any ideas why I could not decode mine properly using python? Thanks again!

**Comment by lbinding** [comment]
@mamuerie Heya, have you tried using a website such as: [https://base64.guru/converter/decode/file](https://base64.guru/converter/decode/file) ? I haven't looked into doing it in Python (although I definitely need to) but this has consistently worked for me being able to play the files. Perhaps see if that works to identify if it is something to do with the saving of the files or the python converting script? 

Becky - thanks for the browser JATOS help - works perfectly when hosted on JATOS :) !!

**Comment by mamuerie** [comment]
@lbinding Thanks for your response. Yes, we have used the same converter to check the files and the files themselves are fine. But in our experiment we will have around 2000 files per participant so I really hope to find a python solution ;-).

**Comment by becky-gilbert** [comment]
@mamuerie so just to clarify: it sounds like your data is saving ok and the problems you're having are in the conversion step. Is that right? Just want to check if there are any problems that originate in the plugin that I need to look into.

As for scripting the conversion, I'm not sure how to do this in Python (sorry!) but here's how I [convert from base 64 to audio in R](https://github.com/jspsych/jsPsych/discussions/796#discussioncomment-33705).

**Comment by mamuerie** [comment]
Hi @becky-gilbert, yes the problem is the conversion step, the original files, and the plugin are working perfect! Thanks for the link, I will try it with R!

**Comment by mamuerie** [comment]
Hi, I just wanted to let you know that I have solved the problem. In case someone has the same problem at some point, here is the solution which worked for me:
The following code creates an webm file from the base64 output
```python
import base64

decodedData = base64.b64decode(b64_str)
webmfile = (outdir + file.split('.')[0] + ".webm")
with open(webmfile, 'wb') as file:
       file.write(decodedData)
```

And the conversion to a wav file worked with ffmpy:

```python
from ffmpy import FFmpeg

ff = FFmpeg(
        executable = 'C:/Program Files/ffmpeg-2020/bin/ffmpeg.exe',
        inputs={file:None},
        outputs = {outfile:'-c:a pcm_f32le'})
ff.cmd
ff.run()
```
After those two steps, I was finally able to read the wav file with matlab or any other program :-).

**Comment by becky-gilbert** [comment]
@mamuerie - great to know that there's a Python option for the base 64 to wav conversion. Thanks very much for sharing!

**Comment by esolitos** [comment]
I am also using the `image-audio-response` plugin by @becky-gilbert on a project for @marta-velnic
It would be great to see this plugin merged into jspsych/jsPsych, so if you need any kind of help to get this merged let me know.

**Comment by becky-gilbert** [comment]
Hi @esolitos, thanks very much for your interest in this plugin and offer to help! I'd like to include it in the next jsPsych release (6.2) but I'm hesitant because I haven't had a chance to do a huge amount of testing. Are you using the most up-to-date version, i.e. this one: https://github.com/becky-gilbert/jsPsych/blob/audio-response/plugins/jspsych-image-audio-response.js? if so, is the mic check working ok? Also, if you're using a custom postprocessing function, it would be good to know if it's working reliably and if you had any problems setting it up. Thanks!

**Comment by esolitos** [comment]
> Are you using the most up-to-date version

Yes. 😊 

>  if so, is the mic check working ok?

Yes from my tests it seems that the check works as expected (Tested on Chrome+Firefox on MacOs, I will need to test it on tablets later). Note that _by design_ it will only work properly when served from a webserver; if the html file is opened in a browser will ask every time the recording starts due to security measures.

> are you custom post processing function?

Not really, but we're still working on the experiment although the  is done, which will then served via JATOS.
Right now I'm not sure if we're going to push the audio as base64 or we will encode them right away in the browser.

> it would be good to know if it's working reliably and if you had any problems setting it up.

We have only been testing some parts, but in general I did not get any kind of trouble, setup was straight forward, as it also works fine with `bjoluc/jspsych-builder`. 😊 

I think it should be specified in the documentation that the "mic permission" will be asked each time when opening the html files directly, considering that a lot of the target users are going to be non-developers.

**Comment by becky-gilbert** [comment]
@esolitos great, thanks very much for the feedback! 

I totally agree about explaining how the mic permissions works in the docs - thanks for pointing this out.

By the way, I also use JATOS and found that using their jatos.uploadResultFile function works really well for saving the data as an audio file - much easier than working with the base64 string. Here's the postprocessing function that I use:
```javascript
function save_file_to_jatos(data) {
  return new Promise(function (resolve) {
    const blob = new Blob(data, { type: 'audio/webm; codecs=opus' });
    // create URL from the audio blob, which is used to replay the audio file (if allow_playback is true)
    let url = URL.createObjectURL(blob);
    // use the trial number as the file name
    var filename = trial_count.toString() + ".webm";
    jatos.uploadResultFile(blob, filename);
    // need to pass an object with the URL and data string to the onRecordingFinish function
    // - url allows the audio to be replayed if playback is true
    // - str is used to either save the audio data as base64 string or save the filename to the jsPsych trial data
    var trial_data = { url: url, str: filename };
    resolve(trial_data);
  });
}
```

**Comment by esolitos** [comment]
Hi Becky, that's quite useful, thanks!

Just a question, since I'm still quite new to jsPsy, before I spend 4h digging into its source: where do you specify such post-process function?

**Comment by becky-gilbert** [comment]
@esolitos the postprocessing function is actually a trial parameter, since we figured that researchers will often want to specify their own way of handling the audio data. So in my study, I include the `save_file_to_jatos` function in my main script, and then use it as a trial parameter like this:
```javascript
var trial = {
  type: 'image-audio-response',
  stimulus: image_white,
  stimulus_duration: 100,
  buffer_length: function () { return buffer_length; },
  allow_playback: false,
  wait_for_mic_approval: false,
  response_ends_trial: true,
  postprocessing: save_file_to_jatos,
  post_trial_gap: function () {
    return jsPsych.randomization.sampleWithoutReplacement(iti_durations, 1);
  }
};
```
The postprocessing function does need to be written in a specific way - sorry about the lack of documentation on this. There is a [description of the postprocessing parameter](https://github.com/becky-gilbert/jsPsych/blob/fbc51f44824023b488ad88d34c0febec2625f8fa/plugins/jspsych-image-audio-response.js#L63) and the [default function](https://github.com/becky-gilbert/jsPsych/blob/fbc51f44824023b488ad88d34c0febec2625f8fa/plugins/jspsych-image-audio-response.js#L37) in the plugin file, which might be helpful

**Comment by kjkendro** [comment]
In a somewhat related question, I'm trying to administer a picture naming task where participants have a maximum of 3 minutes (180000ms) to respond, but I would like them to be able to progress earlier if they finish by pressing the space bar. Right now, I have the following code for each block:
`var mint1_proc = {
      timeline: [
{type: 'html-audio-response',
      stimulus: '<img src="jsPsych/PicNam/Trial_1/1_1.png"></img>'+'<img src="jsPsych/PicNam/Trial_1/1_2.png"></img>'+'<img src="jsPsych/PicNam/Trial_1/1_3.png"></img>'+'<img src="jsPsych/PicNam/Trial_1/1_4.png"></img>'+'<img src="jsPsych/PicNam/Trial_1/1_5.png"></img>'+'<img src="jsPsych/PicNam/Trial_1/1_6.png"></img>'+'<img src="jsPsych/PicNam/Trial_1/1_7.png"></img>'+'<img src="jsPsych/PicNam/Trial_1/1_8.png"></img>'+'<img src="jsPsych/PicNam/Trial_1/1_9.png"></img>'+'<img src="jsPsych/PicNam/Trial_1/1_10.png"></img>',
      buffer_length: 180000,
      allow_playback: false,
      manually_end_recording_key: jsPsych.NO_KEYS,
      stimulus_duration: 180000,
      wait_for_mic_approval: true,
      prompt: 'If you finish early, please wait until the next block appears.'},
      {type: 'html-keyboard-response',
      choices: ['space']
    }
    ]}`

The spacebar is not currently ending the trial; the recording should not end independently of the trial. Any help is greatly appreciated!

**Comment by jackdolgin** [comment]
Sorry to tack on to this thread, but I also had a few questions/ possible suggestions—do you happen to know:

1.  if there's a way in javascript to tell at any given instance what the current input's decibel level is (as a proxy for whether someone is talking)? it could be used to use voice level to determine when to end a recording/trial rather than waiting for a specific amount of time to elapse (I've tried searching for something like this but have had no luck to date)
2) what it means in your code when you comment `start recording with 1 second time between receiving 'ondataavailable' events`, like does that mean the recorder will wait one second before beginning?
3) if one were to end one trial with a recording and immediately begin the next trial with a recording, whether there would be a lag because of how long (if it's long) it takes to postprocess the terminating recording
4) relatedly, if one wanted the timing of onset/offset to be as precise as possible, whether needing to start and end a recording each trial would be as (or even more) accurate as simply recording the entire experiment as one recording and then afterwards offline, say, splicing that recording up by trial?

Thanks so much, this has already been a tremendously useful resource.

**Comment by becky-gilbert** [comment]
@kjkendro we haven't added the option to stop the recording early with a keypress, but that's a good idea. I don't know if I'll have a chance to do this soon. In the meantime, you can do this by adding a function that stops the recording, and calling this function either when the `buffer_length` duration is up, or when a valid key is pressed. Here's the relevant bit of code that stops the recording in the existing plugin, inside the `process_audio` function:
```javascript
setTimeout(function() {
  // this will trigger one final 'ondataavailable' event and set recorder state to 'inactive'
  recorder.stop();
  recorder.wrapUp = true;
}, trial.buffer_length);
```
You could try modifying the plugin like this:
```javascript
// function to initiate the end of recording, either due to a timeout or a keypress
function stop_recording() {
  // check if recorder has been set up and is recording 
  // this is in case a valid key is pressed before recording starts or after it ends 
  if (recorder !== null && recorder.state == "recording") {
    // kill the keyboard listener as soon as recording stops
      jsPsych.pluginAPI.cancelAllKeyboardResponses();
    // the code below will trigger one final 'ondataavailable' event and set recorder state to 'inactive'
    // which will trigger the trial.postprocessing function, followed by onRecordingFinish
    recorder.stop();
    recorder.wrapUp = true;
  }
}

// ... inside the process_audio function ...
setTimeout(function() {
  // stop the recording when buffer_length duration is reached
  stop_recording();
}, trial.buffer_length);

// set up keyboard event listener
var keyboardListener = jsPsych.pluginAPI.getKeyboardResponse({
  callback_function: stop_recording,  // stop the recording when a valid key is pressed
  valid_responses: trial.choices,  // add a choices parameter to the plugin, or use ['space'] instead of trial.choices
  rt_method: 'performance',
  persist: false,
  allow_held_key: false
});
```
But I haven't tested this and not sure if it works.

**Comment by becky-gilbert** [comment]
@jackdolgin No problem! I don't have all of the answers to these questions, so I'm wondering if @mjaquiery wants to add anything, since he wrote the original version of the plugin. Here are my responses:
> 1. if there's a way in javascript to tell at any given instance what the current input's decibel level is (as a proxy for whether someone is talking)? it could be used to use voice level to determine when to end a recording/trial rather than waiting for a specific amount of time to elapse (I've tried searching for something like this but have had no luck to date)

This is possible but might be tricky to set up, because of the audio signal processing involved, and because of differences in background noise and speech levels. If I were to try this, I would use the approach described in [this stackoverflow post](https://stackoverflow.com/questions/62114251/web-audio-api-how-can-i-detect-speech-and-record-until-silence-with-or-without) and [this stackoverflow post](https://stackoverflow.com/a/24518498). This [volume-meter github repo](https://github.com/cwilso/volume-meter) also looks very useful. The idea is to take the audio input stream that comes from the mic and pass that to a WebAudio API `AudioContext`, so that you can use WebAudio API functions to process the data as it comes in. In the audio-response plugins, the audio input stream is passed to the `process_audio` function, so that's where you might want to do the data processing to check levels. Another thing you could do is record some 'silence' at the start of the session, so that you can use that RMS amplitude as your baseline for thresholding.

> 2. what it means in your code when you comment start recording with 1 second time between receiving 'ondataavailable' events, like does that mean the recorder will wait one second before beginning?

So I'm not totally sure what that time argument does, but I do know that the recorder doesn't wait that duration before starting the recording. The [docs for MediaRecorder.start](https://developer.mozilla.org/en-US/docs/Web/API/MediaRecorder/start) say: "this method can optionally be passed a timeslice argument with a value in milliseconds. If this is specified, the _media will be captured in separate chunks of that duration_, rather than the default behavior of recording the media in a single large chunk." So I think this "timeslice" argument is useful for controlling the sampling rate for processing the data. This might be useful if you want to process the data in (close to) real time, as in your example of detecting silence above. 

I recently tested the synchronization between visual display and recording onset for the image-audio-response plugin, and found that the start of the recording is not delayed - it actually seems to always start somewhere around 30-100 ms before the visual display onset. We're working on improving this visual-recording sync.

> 3. if one were to end one trial with a recording and immediately begin the next trial with a recording, whether there would be a lag because of how long (if it's long) it takes to postprocess the terminating recording

I'm not sure. In the plugin code, when the recording stops, that triggers the `trial.postprocessing` function, followed by `onRecordingFinish` and (if `response_ends_trial` is true) `end_trial`. The code in the latter two functions should execute quickly. The default `trial.postprocessing` function uses Promises that are chained together, so this introduces the possibility for delays, but I'm not sure whether the default code is likely to produce delays. Also, this function is a parameter to allow people to use custom code to save the data, so whether or not there are delays will depend on how your custom function is written. If there are delays caused by this function, then that would hold up the chain of other events leading up to ending the trial. This means it could affect the duration between the end of the recording and the start of the next trial, but wouldn't affect the start of the recording on the next trial.

> 4. relatedly, if one wanted the timing of onset/offset to be as precise as possible, whether needing to start and end a recording each trial would be as (or even more) accurate as simply recording the entire experiment as one recording and then afterwards offline, say, splicing that recording up by trial?

Yep I think you could record the whole experiment. As long as you have a way of syncing the start of the recording with a jsPsych timestamp, then you could use the jsPsych timestamps as markers for splicing the recording. But like I said, any delays caused by ending the recordings would happen before the next one starts, and wouldn't affect the sync between recording and stimulus onset for each trial. Also, if you record the whole experiment then I suggest that you use a custom `postprocessing` function to save the .webm file to a server, rather than using the default postprocessing function (because the base64 strings get very very long and will probably cause data storage/transfer problems if you have many and/or long recordings).

**Comment by becky-gilbert** [comment]
@kjkendro oops, I just realized that @hesmall already added in the option to manually end the recording with a key press, with the `manually_end_recording` and `manually_end_recording_key` parameters (thanks @hesmall!). You can see this in [this html-audio-response plugin](https://github.com/becky-gilbert/jsPsych/blob/audio-response/plugins/jspsych-html-audio-response.js) and [this audio-audio-response plugin](https://github.com/becky-gilbert/jsPsych/blob/audio-response/plugins/jspsych-audio-audio-response.js). I'm not sure why I don't have this option in my version of the image-audio-response plugin yet, but I will add it in soon.

**Comment by kjkendro** [comment]
Thank you for the detailed response, @becky-gilbert! I had initially tried to use the `manually_end_recording_key` parameter, with the accepted key press being the spacebar, in two ways:

```javascript
//using 'space'
manually_end_recording_key: ['space'],
//using key code
manually_end_recording_key: ['32'],
```

Both were unsuccessful.  I haven't retrieved the generated audio files yet, but the trial did not progress. I'm not sure if it's a bug or I'm doing something wrong, so I'm going to loop in @hesmall on this, since she had added the parameter.

Regardless, thank you again, and I'll try out the solution detailed in your first comment & report back.

**Comment by becky-gilbert** [comment]
@kjkendro before you spend time modifying the plugin - can you try a key other than the space bar and let me know if that works?

**Comment by kjkendro** [comment]
@becky-gilbert I just tried it with 'j' (and '74'). Like with the spacebar, the trial didn't end/the next trial did not appear.

**Comment by becky-gilbert** [comment]
@kjkendro ok thanks - I'll have a look!

**Comment by mjaquiery** [comment]
@becky-gilbert<https://github.com/becky-gilbert> gave a pretty comprehensive response and I’m not sure I can add anything much! I don’t think I can add much beyond suggesting a bit of experimenting and using the breakpoints in your browser’s debugging tools to figure out what’s going on with the key press issue.
Note that the answers to question 4 and 2 are related: the option for slicing up recordings into smaller packets is there to avoid the problem of large data chunks for long recordings.
Matt

**Comment by kjkendro** [comment]
@becky-gilbert I just got the chance to try out your plugin modification, and it worked like a charm. Thank you so much!

**Comment by jackdolgin** [comment]
@becky-gilbert and @mjaquiery just wanted to say that ultimately what worked for me was using custom (non- js psych) code to record the audio, convert it from webm to wav, and then save the wav in a non-public folder on a server I've been using. I realize not everyone may have access to a server where that's possible, but I basically modeled my code off [this comments on a related blog](https://www.russellgood.com/how-to-convert-audiobuffer-to-audio-file/#comment-453) and its codepen https://codepen.io/rwgood18/pen/wvvyygd (i'm happy to share my script with anyone who asks, i'm keeping the repo private until i complete the study). what's nice is it allows me to record long files, and also i don't have to do anything else to convert to wav files. my task is about 20 minutes split into four blocks, and each block has its own recording. i end each recording after the block ends, which allows me to start saving the resulting ~25 mb .wav files in the background while the task is still running. participants then only need to wait once they've finished the last block and are waiting for that recording to upload, which takes less than a minute anyways (and i use a progress bar to keep them informed).

i didn't get around to testing voice onset and offset in real-time because figuring out the above took long enough, but i'm going to look into the resources you linked before my next go-around. thank you both so much!

**Comment by becky-gilbert** [comment]
@jackdolgin thanks so much for sharing your solution and resources! You're right that not everyone has access to a server that allows uploading arbitrary files, but when that's possible, your solution seems great. So although the jsPsych audio recording plugins need to offer the (non-ideal) text-based/embedded jsPsych data solution, I think we can do a better job of supporting other ways of saving the data, when this is possible. I'm not sure yet how much of this could be done by modifying the plugins vs adding documentation to show how to record and upload wav files - I'll need to think more about this. 

Thanks again for sharing your resources. And if you have any specific suggestions for improving the audio response plugin code, please do feel free to comment here or submit a pull request!

**Comment by lbinding** [comment]
Hi all, 

I'm having a problem with the audio-audio-response plugin downloaded. When I run the program, the audio recording plays, but the visual picture doesn't. Furthermore, the trial doesn't end and presents this error: 

Uncaught (in promise) TypeError: Cannot set property 'onended' of undefined
    at start_trial (jspsych-audio-audio-response.js:239)
    at process_audio (jspsych-audio-audio-response.js:269)

Line 239 corresponds to the else statement. 
`            if(context !== null){
                startTime = context.currentTime;
                source.start(startTime);
            **} else {**
                audio.play();
            }
`

I've had a look at implementing the code @mchlwgnr presents but it doesn't seem to change anything. I was wondering if anyone knew what could be going on? 

Thank you all for the hard work you're doing on this!!!

**Comment by anelleh** [comment]
Thank you all for these amazing plugins!

I'm very new to both jspsych and javascript, but I was wondering what would be the best way to continue the audio recording during the ITI when using the html-audio-response plugin? I don't really want to make a long recording of the whole experiment, but I would like to make sure that all answers are recorded, even if participants answer during the ITI. I tried setting stimulus_duration at 1500 ms and buffer_length at 2000 ms (= 500 ms ITI), which seems to work fine when using Chrome (although the length of the audio files seems to be 1980 ms), but does not work on Firefox. I'm not sure if this is a browser issue, platform issue (currently I'm using Cognition.run to run the experiment) or if I'm using these two parameters wrong?

**Comment by kjkendro** [comment]
@lbinding I believe you would need a different plugin, such as the [html-audio-response plugin](https://github.com/becky-gilbert/jsPsych/blob/audio-response/plugins/jspsych-html-audio-response.js), as I think the audio-audio plugin is for scenarios where both the stimulus and response are audio.

**Comment by kjkendro** [comment]
@anelleh I was able to record response times of up to 3 minutes with this plugin, and @becky-gilbert was so kind as to help me allow participants to proceed via the space bar (modified plugin attached here in a .txt file, so change the extension to .js).

Here's how my variables look:
```javascript
var example = {
      type: 'jspsych-html-audio-response_modified.js',
      stimulus: '<img src="image.png"></img>',
      buffer_length: 180000,
      allow_playback: false,
      manually_end_recording_key: ['space'],
      stimulus_duration: 180000,
      wait_for_mic_approval: true,
      prompt: 'If you finish early, press the spacebar to continue.'
    }
[jspsych-html-audio-response_modified.txt](https://github.com/jspsych/jsPsych/files/6025471/jspsych-html-audio-response_modified.txt)

**Comment by becky-gilbert** [comment]
Hi @anelleh, currently there's no way to extend the recording into the ITI. To do that, we'd have to turn audio recording into a [jsPsych extension](https://www.jspsych.org/extensions/extensions/), so that it can continue collecting data across multiple trials. So for now, your approach makes sense to me: you should be able to set the stimulus duration to be shorter than the buffer length. When you say that this doesn't work in Firefox - what happens exactly? Does the experiment stop working? Does the trial/recording always end after 1500 ms? 

WRT the length of the audio files, I'm not sure what causes this. Hauber et al. posted a [preprint that reports picture naming interference with audio recording based spoken RTs](https://psyarxiv.com/cyvwf/?view_only=229679aa33604aa2a5cb400eab62099), and they also discuss this problem of variable recording lengths. I'll look into this as soon as I get the chance.

**Comment by becky-gilbert** [comment]
@lbinding sorry for the delay getting back to you, and glad to hear that you resolved the issue! Please let us know if you run into any other problems.

**Comment by becky-gilbert** [comment]
Oh and @anelleh another option is to use @kjkendro's version that allows you to manually end the recording with a key press. 

One tip for anyone using the manual key press version - we might want to add in the option of a delay between when the key is pressed and when the recording actually ends, to account for the fact that some participants might press the key while speaking and accidentally cut off part of their response. I'll add this in as a parameter when I get the chance, but for now you can do this by editing the `after_response` function in the plugin file:
```js
var after_response = function(info){

  // after a valid response, the stimulus will have the CSS class 'responded'
  // which can be used to provide visual feedback that a response was recorded
  display_element.querySelector('#jspsych-html-audio-response-stimulus').className += ' responded';

  // only record the first response
  if (response.key == null) {
    response = info;
  }
  // to add a delay after the key press, put the code to end the recording into a setTimeout function
  jsPsych.pluginAPI.setTimeout(function() {
    // this will trigger one final 'ondataavailable' event and set recorder state to 'inactive'
    recorder.stop();
    recorder.wrapUp = true;
  }, 250);  // <-- duration between when the key press is made and when to end the recording, in ms
}
```

**Comment by anelleh** [comment]
@becky-gilbert Thank you for your response! We've tested this quite a few times in Chrome, and sometimes the length of the audio files is 20 ms shorter than the buffer length, but sometimes it's the same as the buffer length. We'll pilot the experiment and see what happens!

Lots of weird things happened when we tested this in Firefox - the length of the audio files was always shorter, and sometimes the trial sequence also changed for some reason (e.g. the test would repeat one trial even though this was not in the script, and turn the last html-audio-response trial into a button-press trial etc.). None of these issues occurred when using Chrome, so we've decided not to use Firefox for now.

**Comment by jackdolgin** [comment]
Following up from before, I've noticed that with the code I've built I also get a slight difference between when the recording begins and my stimuli appear (for me it's been a difference more in the range of 100 - 200 ms, but again different code). I'm curious just conceptually why any lag exists. I'm thinking that for my experiment I might just play a sound at the beginning of recording and use that as a link between display timing and recording timing, but obviously ideally it would be best to avoid that altogether.

**Comment by becky-gilbert** [comment]
Hi @jackdolgin! Thanks for the update. @majorminors and I have been testing lots of different implementations of my version of an audio-response plugin to see if we can get the recording onset to sync with the visual display. I'm using a photodiode and microcontroller to cause a tone that's triggered by the visual display, and then looking at the offset between the start of the audio file and the tone. Is that similar to how you're getting your measurements?

You can find our repo [here](https://github.com/becky-gilbert/audio-response-timing), but it's not up to date. (In case anyone's interested, I'll add my updated versions and testing results to the repo and then edit this comment.) I've tried lots of things, and the best I've done so far is an average of ~30 ms between the start of the recording and the stimuli onset, with a range of ~5-50 ms across trials within a single system/session. That variability seems to big to be useful for measuring spoken response times. I still need to try one other method - fingers crossed it will improve the timing.

I think there are two main categories of error: (1) delays between when the visual display onset is requested and when it actually appears, and (2) when the start of the recording is requested and when it actually starts. The good news is that lots of work has been done on (1) already, and we know that the visual display onset can be within ~1 frame of when it's requested. I don't know how you're putting the stimulus on the screen, but for me the best method has been to add the stimulus to the page with its  CSS `visibility` set to `hidden`, and then show it by changing that to `visible`, which avoids a page reflow. I also suggest using `requestAnimationFrame` to reduce the variability in when the stimulus onset occurs relative to the refresh cycle across trials. I know much less about (2), and if you have suggestions or insight to share, I would love to hear it!

> I'm thinking that for my experiment I might just play a sound at the beginning of recording and use that as a link between display timing and recording timing

Hm, this is an interesting idea, but I don't know how accurate it would be. This still assumes that you've addressed the visual display onset problem as best you can. But on top of that, there's a separate source of error, which is the delay between when you ask the browser to play an audio file and when it actually plays. Both the visual onset and audio onset delays will be variable across systems, since they depend on various software and hardware differences, but they might be variable within systems as well. There's been a bit of discussion about audio-visual synching here: #806. But even if this solution isn't perfect, it might reduce the error enough to produce useable spoken response time measurements.

**Comment by jackdolgin** [comment]
Those are such great points and updates, @becky-gilbert! It seems like you've tested much more carefully than I did- I played sound as my baseline and naively assumed it would emit audio as soon as called, though as you mention that's unlikely. It also sounds like you've nailed down the timing to a manageable range! Depending on the effect sizes that seems like a small enough variance to get meaningful response time results! I've just recorded entire blocks as single recordings, and in my analyses I treat blocks as random effects. But I wish I knew more about the second reason you suggest contributes to delayed RT's. In any event, thank you so much again for your hard work! The progress you've described is quite exciting!

**Comment by anelleh** [comment]
Thank you so much for your hard work on this @becky-gilbert, amazing how thoroughly you've tested the audio-visual timing!

Regarding the duration of the audio files, I went through all the audio files for our pilot (144 files for each participant), and all the files are either 2100 ms long (which is the set buffer length) or 2040 ms long. The percentage of audio files that are of the correct duration (2100 ms) varies significantly across participants, from approx. 20 to 80 %. Do you know if there's any way to tell if the end or the beginning of the shorter audio files has been cut? Looking at your repo, so far I've been using the "standard" method to measure response times (using the start of the audio file as the stim onset time, and calculating RT as the delay between the start of the audio file and audio response onset), but would it maybe help to change to "start recording first" method (start audio recording first and get a timestamp, then display the image and get timestamp)? Though I'm not really sure I know how to modify the html-audio-response-plugin to use the `requestAnimationFrame`.

**Comment by Patlane** [comment]
This is really great work. I'm so fascinated by the flexibility of jsPsych! I'm curious.
For the sound data, there is a save_file_to_jatos() function.  I'd love to collect the webm-data (that is audio files) with cognition.run. How I should go about it? Do I need a function to save audio on congnition.run?

**Comment by myrtevos** [comment]
I've recently written an audio response recording plugin for a specific usecase (eliciting words from dialect speakers, for which we need to allow informants to manually start, stop, and re-record audio). It's a hybrid between Becky's html-audio-response plugin and Michael Weiss (m-w-w)'s record-audio-JATOS plugin (thanks to both!) and uploads audio as a result file on JATOS. 

I still need to file down some of the rough edges (it relies on jQuery and an associated CSS file, I haven't tested it super thoroughly, and I want to make rerecording an optional parameter), [so right now it's half-buried in a tutorial repo](https://github.com/myrtevos/magnus-jspsych), but I thought I'd post it here in case it's useful to anyone else :)

**Comment by zhangchi199133** [comment]
Dear @becky-gilbert and @hesmall, thanks for the amazing plugins. It helps me greatly. I am now using the html-audio-response plugin for its feature of manually ending the recording with a key press. I was wondering whether I can apply the same feature to a button press? If so, how should I modify the plugin? Many thanks!

**Comment by jodeleeuw** [comment]
Hi everyone who is following this thread,

We've now added the [initialize-microphone](https://www.jspsych.org/latest/plugins/initialize-microphone/) and [html-audio-response](https://www.jspsych.org/latest/plugins/html-audio-response/) plugins to our main repo. These don't implement all of the features that have been discussed in this thread, but they do cover many of them. I'm going to close this thread for now since the main objective is complete, but we'd love feedback about the audio recording features and please feel free to open up a new issue for any specific requests or problems that arise using those plugins.

**Comment by nelsonroque** [comment]
Hi @jodeleeuw - thank you so much for making this possible! Should this base64 data be converted to webm audio or wav audio? I tried converting to wav and it works, but the metadata is corrupt, resulting in downstream processing tasks (e.g., with OpenSMILE) complaining about a missing header RIFF ID. Thank you for any guidance!
